### PR TITLE
v0.0.129 — WASM call translator major correctness fixes (#475 PR 2 of 2)

### LINKED ISSUE #475: Pre-existing bugs in WASM call translators (surfaced by #474 review)
During review of PR #474 (calls.py decomposition — pure code motion), CodeRabbit identified 10 pre-existing bugs in the WASM call translator code. These were not introduced by the refactor; they existed in `vera/wasm/calls.py` before the split. Filing them here so they're tracked but don't contaminate the mechanical refactor PR.

## Findings

| # | Severity | Location (post-refactor) | Description |
|---|---|---|---|
| 1 | 🔴 Critical | `calls_handlers.py` `_translate_handle_exn` | Catch-arm result type inference only checks when `clause.body` is `ast.Block`; expression-bodied handlers leave `result_wt = None`, omitting the `(result ...)` annotation in emitted WAT. |
| 2 | 🔴 Critical | `calls_strings.py` `_translate_string_slice` | Slice bounds narrowed with `i32.wrap_i64` before clamping — large positive i64 indices wrap to negative i32 values, then get clamped to 0. Should clamp in i64 first, wrap after. |
| 3 | 🔴 Critical | `calls_strings.py` `_translate_char_code` | No bounds check on index before `i32.load8_u` — out-of-range index reads arbitrary memory. |
| 4 | 🟠 Major | `calls_arrays.py` `_translate_array_slice` | Same i64→i32 narrowing issue as #2 (both start and end indices). |
| 5 | 🟠 Major | `calls_containers.py` `_map_wasm_tag` | Fallback routes every non-primitive/non-String type to `"b"` (i32), but `Array<T>` lowers to i32_pair. Breaks `map_insert`/`map_values` for `Map<K, Array<T>>`. Should reject Array values with a clear error. |
| 6 | 🟠 Major | `calls_encoding.py` `_translate_url_parse` / `_translate_url_join` | `url_parse` discards `has_auth` and delimiter-presence info, so `url_join(url_parse("http:path"))` round-trips as `http://path`. Need to preserve `scheme_has_authority`, `query_delimiter_present`, `fragment_delimiter_present` flags. |
| 7 | 🟠 Major | `calls_encoding.py` `_translate_base64_decode` | Only counts trailing `=` into pad but still treats `=` as value 0 in the main loop. `=` in non-padding positions should be rejected. |
| 8 | 🟠 Major | `calls_parsing.py` `_translate_parse_nat` / `_translate_parse_int` | Space-skip loop runs unconditionally, so embedded spaces are silently accepted: `"12 34"` parses as 1234, `"- 5"` parses as -5. Should only skip spaces before the first digit/sign. |
| 9 | 🟠 Major | `calls_strings.py` `_translate_to_string` | Integer negation of `INT64_MIN` (-9223372036854775808) overflows (no positive representation fits in i64), producing wrong output. Need INT64_MIN special case or unsigned-magnitude conversion. |
| 10 | 🟠 Major | `calls_strings.py` `_translate_float_to_string` | Rounded fractional part can equal 1_000_000 (e.g., 1.9999995), dropping the carry and producing `1.0` instead of `2.0`. Need carry detection after `i64.trunc_f64_s`. |

## Verification

Each finding references code that was moved by #474 without modification. To confirm pre-existence, compare the cited lines against `git show main:vera/wasm/calls.py` — the bug patterns match exactly.

## Priority

The three Critical findings (1, 2, 3) should be addressed first. 8 (parse_nat/int embedded spaces) is almost certainly user-visible. The others are edge cases but represent real semantic gaps.

Filed from PR #474 review.

### PR DESCRIPTION
**Pull request by aallan**
## Summary
- Ships the seven Major correctness findings from CodeRabbit's review of PR #474's `calls.py` decomposition (PR 2 of 2 for **[#475](https://github.com/aallan/vera/issues/475)**). PR 1 (v0.0.128) shipped the three Critical safety fixes and prematurely auto-closed #475 via its commit subject; the issue has been reopened so this PR can close it correctly.
- All seven fixes are small, independent translator changes in `vera/wasm/calls_*.py`. No grammar, parser, or type-checker changes.
- 22 new test cases across 7 new test classes; no regressions in the 3,658 pre-existing tests.

## The seven fixes

1. **`array_slice` clamps in i64 before wrapping** (`calls_arrays.py`) — same shape as the v0.0.128 `string_slice` fix. Pre-fix, large i64 indices wrapped to small in-range-looking i32 values and copied past the array. Post-fix uses the cross-mixin `_clamp_i64_to_range_then_wrap` helper (shared via Python MRO with `CallsStringsMixin`).
2. **`Map<K, Array<T>>` rejected at codegen** (`calls_containers.py`) — `_map_wasm_tag` return type changed from `str` to `str | None`; 11 call sites guard for `None`. Pre-fix the type system held but codegen produced compromised binaries treating Array values as opaque pointers.
3. **`url_parse` / `url_join` round-trip preserves URL shape** (`calls_encoding.py`) — flag bits packed at struct offset 44 with an explicit-mode sentinel (bit 3). When sentinel is set, `url_join` uses bits 0-2 for `has_auth` / `has_query` / `has_frag`; when clear (data-constructor case), falls back to the legacy `len > 0` heuristic so existing direct-construction tests still pass.
4. **`base64_decode` rejects `=` outside the padding region** (`calls_encoding.py`) — RFC 4648-compliant position check (`idx >= slen - pad`). Pre-fix `AB=C` decoded silently with the embedded `=` treated as zero bits.
5. **`parse_nat` / `parse_int` reject embedded spaces** (`calls_parsing.py`) — leading whitespace still trimmed; trailing whitespace still allowed via a separate post-loop tail-validation block; embedded space (between digits) breaks the digit loop and the tail-validator's "every remaining byte must be a space" check fires.
6. **`int_to_string(INT64_MIN)` correct** (`calls_strings.py`) — loop break uses unsigned `i64.eqz` after `i64.div_u` / `i64.rem_u` digit extraction. Pre-fix the signed `i64.le_s 0` break terminated immediately because `-INT64_MIN` overflows back to `INT64_MIN`.
7. **`float_to_string` handles fraction-rounding carry** (`calls_strings.py`) — `frac_val = round((f - floor(f)) * 1_000_000)` is computed *before* the integer part is emitted; if `frac_val == 1_000_000` we increment the integer and reset `frac_val` to 0. Pre-fix `1.9999996` printed as `1.0` instead of `2.0`.

## Tests added

| Test class | Tests | Coverage |
|---|---|---|
| `TestArraySliceClamp475` | 4 | normal, negative-start, end-beyond-length, i64-overflow |
| `TestMapArrayValueRejected475` | 1 | `Map<Nat, Array<Nat>>` codegen rejection |
| `TestUrlParseJoinRoundTrip475` | 5 | `http:path`, full URLs, query-with-body, empty `?`, empty `#` |
| `TestBase64DecodePadding475` | 3 | valid no-padding, valid one-pad, misplaced `=` rejected |
| `TestParseEmbeddedSpaces475` | 4 | normal, leading-space-OK, embedded-space-rejected (nat + int) |
| `TestToStringInt64Min475` | 2 | INT64_MIN, negative-basic |
| `TestFloatToStringCarry475` | 3 | carry-propagates, normal, full-six-decimals |

## Test plan

- [x] All 22 new tests pass
- [x] Pre-existing 3,658 tests pass (no regressions)
- [x] `mypy vera/` clean
- [x] `python scripts/check_conformance.py` — all 82 conformance programs pass
- [x] `python scripts/check_examples.py` — all 33 examples pass check + verify
- [x] `python scripts/check_doc_counts.py` clean
- [x] `python scripts/check_version_sync.py` — v0.0.129 consistent across `pyproject.toml`, `vera/__init__.py`, `docs/index.html`, `README.md`
- [ ] CI green

Closes #475

## Release Notes – v0.0.129

* **Bug Fixes**
  * Fixed array slice handling with large indices
  * Corrected URL parsing and joining to preserve delimiters consistently
  * Strengthened Base64 decoding to enforce valid padding placement
  * Enhanced number parsing to reject embedded spaces
  * Fixed edge cases in integer and floating-point string conversion

* **Tests**
  * Added regression tests for recent WASM translator correctness fixes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,18 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 
 ## [Unreleased]
 
+## [0.0.129] - 2026-04-29
```
[the remark this answers is not in the document]
Comment (aallan) [reply CHANGELOG.md]: @coderabbitai Fixed in 30f721b — the v0.0.129 header is now 2026-05-05 (same calendar day as v0.0.128, since today is 2026-05-05; the original 2026-04-29 was a copy-paste error from the v0.0.127 line).

File: HISTORY.md [hunk 2]
```diff
@@ -263,6 +263,7 @@ Stage 11 shifts focus from evaluation infrastructure to the standard library and
 | v0.0.126 | 28 Apr | **Tail-recursive iteration runs in constant stack space** ([#517](https://github.com/aallan/vera/issues/517)). |
 | v0.0.127 | 29 Apr | **`@Nat` subtraction soundness hole closed** ([#520](https://github.com/aallan/vera/issues/520)). |
 | v0.0.128 | 5 May | **WASM call translator critical safety fixes** ([#475](https://github.com/aallan/vera/issues/475)). |
+| v0.0.129 | 29 Apr | **WASM call translator major correctness fixes — closes [#475](https://github.com/aallan/vera/issues/475)**. |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Fixed in 30f721b — the v0.0.129 row in the Stage 11 table now reads `5 May` (same fix family as the CHANGELOG date — copy-paste error from a prior v0.0.127 row).

File: tests/test_codegen.py [hunk 3]
```diff
@@ -14430,3 +14430,419 @@ def test_last_valid_index(self) -> None:
...
+{
+  let m = map_set(Map.empty(), 1, [1, 2, 3]);
+  IO.print(nat_to_string(array_length(map_get(@Map.0, 1))))
+}
+"""
+        # Compiler should fail — either type-check rejects the
+        # nested Array value, or codegen returns None for the
+        # unsupported map flavour.
+        with pytest.raises((Exception,)):
+            _compile_ok(src)
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py]: @coderabbitai Fixed in 30f721b — the test now uses valid Vera (`map_new()` instead of `Map.empty()`, bare type names `<Nat, Array<Nat>>` instead of `<@Nat, @Array<@Nat>>`, no spurious `@Map.0` reference). It now asserts the *exact* codegen-skip path: `_compile_ok` succeeds (no error diagnostics), but `main` is missing from `result.exports` and an E602 'unsupported' warning is emitted. Renamed to `test_compile_skips_function_for_map_of_array` to match the new assertion shape.

File: vera/wasm/calls_containers.py [hunk 4]
```diff
@@ -133,18 +133,37 @@ def _translate_decimal_round(
...
         if vera_type == "String":
             return "s"   # i32_pair
+        # Array<T> values lower to i32_pair too, but no Map host
+        # import currently handles that shape.  Reject so the caller
+        # bails to "function skipped" rather than emitting a
+        # signature-mismatched import.
+        if vera_type is not None and vera_type.startswith("Array"):
+            return None
         # Bool, Byte, ADTs, Map handles → i32
         return "b"
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_containers.py:183]: @coderabbitai Partially addressed in 30f721b. I tried the strict `if vera_type is None: return None` rejection you suggested and it broke 3 pre-existing tests (`test_set_empty_to_array`, `test_map_keys_in_if_expr`, `test_set_to_array_in_if_expr`). These tests rely on the `None → "b"` fall-through specifically because the element type is *legitimately* unknown for empty collections (`set_new()` with no elements added) but the host import works fine — no element value ever flows through the i32 slot.

Kept the explicit `Array<...>` rejection (which is the load-bearing fix for `Map<K, Array<T>>` mis-tagging) and reverted the strict None rejection. Updated the docstring to document why the fall-through is intentional and which tests pin it. The mis-tagging hole the original fix targets is still closed: any *real* (non-None, non-Array) type that fails inference would have to be a brand-new value shape we don't recognize, which is a separate concern from empty-collection inference.

If you have a more surgical fix in mind that distinguishes "genuinely empty collection" from "unknown non-empty type", happy to consider it as a follow-up.

File: vera/wasm/calls_encoding.py [hunk 5]
```diff
@@ -1851,31 +1904,95 @@ def _translate_url_join(
...
         ins.append(f"local.get {s_len}")
         ins.append("i32.const 0")
         ins.append("i32.gt_u")
         ins.append("if")
         ins.append(f"  local.get {total}")
-        ins.append("  i32.const 3")
+        ins.append("  i32.const 1")
         ins.append("  i32.add")
         ins.append(f"  local.set {total}")
         ins.append("end")
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_encoding.py:1978]: @coderabbitai Skipped in this PR; filed as [#568](https://github.com/aallan/vera/issues/568) and added to KNOWN_ISSUES.md (per project convention that PR-surfaced bugs need either a fix or a tracking issue). Verified pre-existing on `main` at 448924c (`vera/wasm/calls_encoding.py` decomposition commit) — `url_parse(":foo") + url_join` produces `foo` both before and after PR 2's seven Major fixes. PR 2's flag-bit work targeted `has_auth`/`has_query`/`has_frag` (#475 finding 6); a `has_colon` flag for the leading-colon case is a fourth flag-bit dimension that would expand scope.

#568 documents both viable fixes (RFC 3986-compliant rejection in `url_parse` for `colon_pos == 0`, vs. extending the offset-44 packed flag word with a `has_colon` bit) for the follow-up PR.

**Comment by aallan** [comment]
@coderabbitai Closing out the outside-of-diff comments from both rounds:

**Round 1 — 3 outside-of-diff:**

1. **HISTORY.md line 306 — `128 tagged releases` → `129`** ✅ Fixed in 30f721b. Also synchronised the `By the numbers` table column header from `v0.0.128 (5 May)` → `v0.0.129 (5 May)` and bumped the test-count cell from 3,658 → 3,680 to match.

2. **ROADMAP.md line 282 — `128 tagged releases (as of v0.0.128), 3,658 tests` → `129 (as of v0.0.129), 3,680 tests`** ✅ Fixed in 30f721b. Also fixed the opening line 3 (`Vera v0.0.128 delivers...` → `v0.0.129`).

3. **calls_parsing.py 80-154 — `is_space_byte` helper refactor across `parse_nat` / `parse_int` / `parse_bool` / `parse_float64`** ⏭ Skipped. Valid code-quality suggestion, but it touches 4 translator functions for a stylistic consolidation when the user instruction was 'keep changes minimal'. Defer to a follow-up code-quality PR; the duplicated `i32.const 32 / i32.eq` predicate is harmless and the surrounding loop structures differ enough that a clean shared helper signature would need its own design pass.

**Round 2 — 1 outside-of-diff + 1 duplicate:**

4. **tests/test_codegen.py 14435-14864 — use `_check_ok`/`_check_err`/`_verify_ok`/`_verify_err` instead of `_run`/`_run_io`/`_compile_ok`** ⏭ Skipped — finding is misapplied to this file. The CLAUDE.md helper convention applies to `test_checker.py` and `test_verifier.py` (where they're the standard pattern). `tests/test_codegen.py` is for codegen-execution tests and uses `_run` / `_run_io` / `_compile_ok` throughout its 1,033 existing tests — including the v0.0.128 PR 1 tests (`TestStringSliceClampBefore475`, `TestCharCodeBoundsCheck475`, `TestExpressionBodiedExnHandler475`) that you reviewed and approved. Migrating these 7 new test classes to the checker/verifier pattern would diverge from the file's established convention and break the assertion shape (`_check_ok` doesn't compile/execute; we need exit code, stdout, exports, and trap-vs-success).

5. **calls_containers.py 170-183 (duplicate of round-1 finding 3188163502)** — Already replied inline at [discussion_r3188281905](https://github.com/aallan/vera/pull/567#discussion_r3188281905); summary: kept the `None → "b"` fall-through because strict `None`-rejection broke 3 pre-existing empty-collection inference tests. Explicit `Array<...>` rejection is retained (the load-bearing fix for `Map<K, Array<T>>` mis-tagging). If you have a more surgical caller-side classifier in mind that distinguishes 'genuinely empty collection' from 'unknown non-empty type', happy to spike it as a follow-up.

All fixes validated: mypy clean, 3,680 tests pass, 82 conformance programs pass, 33 examples pass.
